File: Csharp/RT_CS/RT_CS/utility/Renderer.cs
using System;
using System.Collections.Generic;
using System.Text;
using RT_CS.sceneObjects;
using RT_CS.Cameras;
using RT_CS.utility;
using RT_CS.shapes;
using RT_CS.utility.IO;

namespace RT_CS.utility
{
    class Renderer
    {
        private SceneDatabase database; 
        public Renderer (SceneDatabase scd)
        {
            database = scd;
        }
        private IntersectionRecord RunIntersection (Camera cam, List<Shape> shapes, int pX, int pY)
        {
            IntersectionRecord inter = new IntersectionRecord();
            Ray myRay = cam.Project(pX, pY);
            foreach (Shape s in shapes)
            {
                s.Intersect(myRay, inter);
            }
            return inter;
        }
        public Vector3[,] Render ()
        {
            int width = database.MyScene.Width;
            int height = database.MyScene.Height;
            Camera c1 = database.MyCamera;
            List<Shape> myShapes = database.MyShapes;
            Vector3[,] image = new Vector3[height,width];
            for (int y = 0; y<height; y++)
            {

                for (int x = 0; x<width; x++)
                {
                    IntersectionRecord record = RunIntersection(c1, myShapes, x, y);
                    Vector3 pxcolour = new Vector3();
                    if (record.Hit)
                    {
                        pxcolour.x = 1;
                        pxcolour.y = 1;
                        pxcolour.z = 1;
                    }
                    else
                    {
                        pxcolour = database.MyScene.Background;
                    }
                    image[y, x] = pxcolour;
                }
            }
            return image;
        }
    }
}

File: cpp/RT_CPP/RT_CPP/renderer.cpp
#include "renderer.h"
#include "camera.h"
#include "ray.h"

using namespace cameras;
using namespace utility;



renderer::renderer(scene_database& db) : db(db)
{

}

vector3** renderer::render()
{
	int width = db.get_scene().get_width();
	int height = db.get_scene().get_height();
	camera& cam = db.get_camera();
	vector<shape*> shapes = db.get_shapes();
	vector3** finalImages = new vector3*[height];
	for (int y = 0; y < height; y++)
	{
		finalImages[y] = new vector3[width];
		for (int x = 0; x < width; x++)
		{
			intersection_record* rec = run_intersection(cam, shapes, x, y);
			vector3 pxColor;
			if (rec->hit)
			{
				shape* shape = db.get_shape_by_id(rec->shape_id);
				material* material = db.get_material_by_id(shape->get_material_id());
				pxColor = shape->paint(*rec, db.get_lights(),  db.get_scene().get_ambient_light(), cam.get_view_direction(),db.get_scene().get_generation(),
					&shapes,&db.get_materials());
			}
			else
			{
				pxColor = db.get_scene().get_background();
			}
			finalImages[y][x] = pxColor;
			delete rec;
		}

	}
	return finalImages;
}


intersection_record* renderer::run_intersection(camera const& cam, vector<shape*> const& shapes, int pX, int pY)
{
	intersection_record* rec = new intersection_record;
	ray r = cam.project(pX, pY);
	for (shape* s : shapes)
	{
		s->intersect(r, *rec);
	}
	return rec;

}